### ISSUE TITLE
The Impossibility of Trisecting the Angle and Doubling the Cube (metamath 100 #8)

### ISSUE THREAD

**Issue by tirix** [metamath/set.mm#4246]
An elegant proof would certainly rely on Galois Theory.

Such a proof is provided Chapter 7.4 of [[Ian Stweart - Galois Theory]](https://eclass.uoa.gr/modules/document/file.php/MATH594/Stewart%20Galois%204th%20edition.pdf), p.99, theorems 7.13 and 7.14.

**Comment by savask** [comment]
Maybe worth mentioning that the proof doesn't require a _great deal_ of Galois theory, in fact the only crucial bit seems to be that field extensions are multiplicative, and that is already proved in set.mm in [extdgmul](https://us.metamath.org/mpeuni/extdgmul.html). The fact that the cubic root of 2 gives us a field extension of degree 3 can be proved by elementary means, by mimicking the argument about minimal polynomials.

Probably the most laborious part would be to define constructible numbers and prove that they give field extensions of degree a power of 2. I took a look at how other formalization systems do this, and both HOL Light and Isabelle go with the [geometrical definition](https://en.wikipedia.org/wiki/Constructible_number#Geometrically_constructible_points). That is, we start with two points (if we identify points on the plane with vectors, then usually `(0, 0)` and `(1, 0)`) and call them constructible by default. Then:
1) if points A, B, C, D are constructible, then the intersection of proper non-parallel lines AB and CD is again constructible,
2) if points A, B, C, D, E, F are constructible, then the intersection points of a proper line AB with a circle with center at C and radius |EF| are again constructible,
3) finally, if A, B, C, D are constructible, then the intersection points of two distinct circles with centers A, C and radii |AB|, |CD|, respectively, are also constructible.

I think it would be beneficial to model constructible numbers in set.mm as a subset of complex numbers, as we do in section "Theorems of Pythagoras, isosceles triangles, and intersecting chords" of set.mm. Geometry of `CC` is more developed in set.mm than the Tarskian geometry, and this gives an additional benefit that we may identify points on the plane with points of `CC` and hence a real number is constructible if the corresponding complex number (i.e. point) is constructible.

I unfortunately don't know how recursive definitions work in set.mm, so I cannot draft a preliminary definition of constructible (complex) numbers, but it seems that we should start with 0 and 1 as being constructible by default and properties 1)-3) can be easily stated with standard operations with complex numbers, for example, one can use something in the spirit of [sigarcol](https://us.metamath.org/mpeuni/sigarcol.html) to test if points lie on the same line.

The next step would be to show that construcible numbers form a field (again, I think the fact that we would be able to use `+` and `x` on complex numbers should simplify the proof), and then perhaps the most difficult step would be to show that if a number is constructible then the corresponding field extension degree is a power of 2. I don't know if set.mm has notation for a field generated by a subfield and some element, but more importantly, we would have to prove that intersections of circles and lines require only quadratic field extensions.

I would be glad to participate and discuss the details if someone is willing to try formalizing some of the definitions and intermediate steps required.

**Comment by icecream17** [comment]
Something like `( rec ( ( s e. _V |-> AllPointsConstructibleFrom_s ) , { <. 0 , 0 >. , <. 1 , 0 >. } ) " _om )` would work

**Comment by savask** [comment]
I tried giving a definition of constructible points (and numbers) using icecream17's `rec` idea:
```
$c Constr $.
cconstr $a class Constr $.

df-constr $a |- Constr = ( rec ( ( s e. _V |-> { x e. CC |
 ( ( E. a e. s E. b e. s E. c e. s E. d e. s E. t e. RR E. r e. RR ( x = ( a + ( t x. ( b - a ) ) ) /\ x = ( c + ( r x. ( d - c ) ) ) /\ -. ( Im ` ( ( * ` ( b - a ) ) x. ( d - c ) ) ) = 0 ) \/
 E. a e. s E. b e. s E. c e. s E. d e. s E. e e. s E. f e. s E. t e. RR ( x = ( a + ( t x. ( b - a ) ) ) /\ ( abs ` ( x - c ) ) = ( abs ` ( e - f ) ) ) ) \/
 E. a e. s E. b e. s E. c e. s E. d e. s E. e e. s E. f e. s E. t e. RR ( -. a = d /\ ( abs ` ( x - a ) ) = ( abs ` ( b - c ) ) /\ ( abs ` ( x - d ) ) = ( abs ` ( e - f ) ) ) )
 } ) , { 0 , 1 } ) " _om ) $.
```
Here constructible points are understood as a subset of complex numbers `CC`. We start with `0` and `1` being constructible by default, and the first line says that the intersection of two not-parallel constructible lines is constructible (notice how the expression with `Im` checks for collinearity, it was used in [sigarcol](https://us.metamath.org/mpeuni/sigarcol.html) for the same purpose), second line is for the intersection of a line and a circle, and the third line is for the intersection of two circles with different centers. This is pretty much the same definition as Isabelle or HOL Light use, but inside the complex plane.

I guess one could similarly give a more general definition of "points constructible from some set", but I don't think this is important enough to complicate the definition.

What do you think? Does this definition look right? If we agree on this definition, we could try writing down intermediate statements required to prove the metamath 100 question. Norman [suggested](https://groups.google.com/g/metamath/c/5vbEj2vnvAg/m/eJ5qnyS5BQAJ) formalizing statements (without proofs) of metamath 100 theorems in the past, so this could be a good idea.

**Comment by tirix** [comment]
Nice initiative!

It was not immediately clear to me, but the actual condition to express
is that the intersecting lines and circles should not be the same.
Otherwise the whole line/circle gets added, and we end up with too many
constructible points. If there is no intersection, no point gets added.

Your condition expressed with `|Im|` and the `|-. a = d|` condition are
both sufficient conditions for the line, resp. circle to be distinct,
and it does not require an additional bound variable, so it looks
correct, and clever!

I don't think we have a notation yet for a field generated by a subfield
and and some additional element (like e.g. ℚ[√2]), that would certainly
be a nice addition, and I'd be interested to contribute!

**Comment by savask** [comment]
>  so it looks correct

Nice, so at very least we can express the statement of the theorem now (cubic root of 2 and cos(pi/9) don't lie in `Constr`).

> I don't think we have a notation yet for a field generated by a subfield and and some additional element (like e.g. ℚ[√2]), that would certainly be a nice addition, and I'd be interested to contribute!

That's an interesting question whether we really need a full-fledged definition for the proof. We certainly want to work inside `CC`, so let's say `s` is a subfield of `CC` (viewed as a subset), and `c e. CC` is such that `( c x. c ) e. s`, so a square root of some element from `s`. Then
```
{ x e. CC | E. a e. s E. b e. s x = ( a + ( b x. c ) ) }
```
is also a subfield of `CC`. We want to prove that any constuctible number can be obtained by a finite series of field extensions like that. Maybe we could define "algebraically constructible numbers" as
```
$c ConstrAlg $.
cconstralg $a class ConstrAlg $.

df-constralg $a |- ConstrAlg = ( rec ( ( s e. _V |-> { x e. CC | E. a e. s E. b e. s E. c e. CC ( ( c x. c ) e. s /\ x = ( a + ( b x. c ) ) ) } ) , QQ ) " _om ) $.
```
and then we need to prove that `Constr C_ ContrAlg` (they are equal, of course, but we require only inclusion for impossibility proofs). Now, I'm not sure if it's straightforward enough to relate quadratic field extensions inside `CC` I described earlier with this definition of `ConstrAlg`, so your thoughts are appreciated. Certainly the recursive operation in the definition of `ConstrAlg` doesn't always give a field, so one would need some sort of proof by induction, showing that if a number can be obtained from `QQ` by `k` iterations of this operation, then it lies in a finite field extension of `QQ` of index a power of 2.

**Comment by tirix** [comment]
I believe this approach would work, and we don't really need the full-fledged definition of a generated field for the proof. But it would be sooo nice to use this opportunity to formalize more of the Galois Theory and have some basic bricks for e.g. the insolvability of the quintic...

**Comment by tirix** [comment]
My thought was defining something like:
```
df-fldgen $a |- fldGen = ( f e. _V |-> ( e e. ~P ( Base ` f ), s e. ~P ( Base ` f ) |-> |^| { a | ( s C_ a /\ f /FldExt a /\ e /FldExt a ) } ) )
```
Then given a universe field `F`, a subfield `E` and a generating set `S`, ``( E ( fldGen ` F ) S )`` would be the field extention of `E` generated by `S` in `F`. E.g.  we could write ``( ( CCfld |`s QQ ) ( fldGen ` CCfld ) { ( sqrt ` 2 ) } )``.

**Comment by savask** [comment]
> But it would be sooo nice to use this opportunity to formalize more of the Galois Theory and have some basic bricks for e.g. the insolvability of the quintic...

I agree, then let's see if we can formulate it in Galois theory terms.

> My thought was defining something like: ...

Any reason why we need a subfield `E` at all? Why not define `fldGen` for a universe field `F` and a subset `S`, so `F fldGen S` will be the field generated by `S` inside `F`. In theorems where we extend a subfield `E` with elements from `S` we could use an idiom ```F fldGen ( ( Base ` E ) u. S )```.

Also, in your definition you intersect fields with `|^|` operation, is it going to work? I don't understand extensible structures in set.mm very well, but I would assume that one first has to intersect sets of the form ```( Base ` a )```, and then repackage the result back into a structure.

**Comment by tirix** [comment]
> In your definition you intersect fields with `|^|` operation, is it going to work? I don't understand extensible structures in set.mm very well, but I would assume that one first has to intersect sets of the form `` ( Base ` a ) ``, and then repackage the result back into a structure.
>
You're right, this would not work. For an intersection-based definition, the resulting object has to be a subset of the base. We can then recover a field by using the structure restriction operator: ``( F |`s X )``

> Any reason why we need a subfield `E` at all? Why not define `fldGen` for a universe field `F` and a subset `S`, so `F fldGen S` will be the field generated by `S` inside `F`. In theorems where we extend a subfield `E` with elements from `S` we could use an idiom `` F fldGen ( ( Base ` E ) u. S ) ``.
> 
That's a good remark, this would simplify the definition to really "a field generated by a set" without even mentioning field extension.
A definition could be:
```
df-fldgen $a |- fldGen = ( f e. _V , s e. ~P ( Base ` f )  |-> |^| { a e. ( ( SubRing ` f ) i^i Field ) | s C_ a } ) )
```
This resembles the definition of sigma algebra generated by a set ([`df-sigagen`](https://us.metamath.org/mpeuni/df-sigagen.html)), or that of the ring spanned by a set ([`df-rgspn`](https://us.metamath.org/mpeuni/df-rgspn.html))

**Comment by tirix** [comment]
See #4553 for my concrete proposal.

**Comment by tirix** [comment]
I think another thing we would need to formalize is the idea of "tower" of field extensions. Ideally, we should find a way that is generic enough so that it could also be used for [towers of normal subgroups](https://en.wikipedia.org/wiki/Subgroup_series#Normal_series.2C_subnormal_series), as these are useful for the definition of solvable groups.

I believe the best is to see such towers as [chains](https://en.wikipedia.org/wiki/Lattice_(order)#Jordan%E2%80%93Dedekind_chain_condition), i.e. see the subfield (resp. normal subgroup) relation as an order relation, so that the chains are subsets where the order is total.

We already have definitions for Posets and Tosets, so I think we could prove some generic facts in the settings of order theory, and use them for such chains (I don't think we need a lot of them)

**Comment by tirix** [comment]
> I believe the best is to see such towers as [chains](https://en.wikipedia.org/wiki/Lattice_(order)#Jordan%E2%80%93Dedekind_chain_condition), i.e. see the subfield (resp. normal subgroup) relation as an order relation, so that the chains are subsets where the order is total.

(That's actually what I had in mind when I defined the Field Extension as a _relation_)

**Comment by savask** [comment]
I agree that we may need towers for general Galois theory, and it looks very logical to model them as chains in posets. Nevertheless, I suggest we choose the simplest path towards trisecting the angle and doubling the cube, so we don't spread ourselves thin :-) For these two theorems, one doesn't need to do a detour into group theory, and I think one doesn't need towers of fields/subgroups.

Since we are using field extensions now, we could give the following definition of "algebraic" constructible numbers:
```
df-constralg $a |- ConstrAlg = { x e. CC | E. k e. NN E. f e. Field ( ( CCfld /FldExt f /\ f /FldExt ( CCfld |`s QQ ) ) /\ ( f [:] ( CCfld |`s QQ ) ) = ( 2 ^ k ) /\ x e. ( Base ` f ) ) } $.
```
I.e. a number is constructible, if it lies in some finite field extension of `QQ` of degree a power of 2. Two big things to prove would be that `Constr C_ ConstrAlg` and `-. ( 2 ^c ( 1 / 3 ) ) e. ConstrAlg` (and similarly for trisecting the angle). It seems that to prove the first part we'll have to use induction over the definition of `Constr`, for example, if two nonparallel lines pass through points which lie in fields F1, F2, F3, F4, and |F_i : Q|is a power of 2, then the composite field K generated by F_i is also an extension of Q of degree a power of 2, and the intersection point lies in K (and thus in `ConstrAlg`). Here we will definitely use `fldGen`, and `extdgmul`. We'll also need a theorem saying that adjoining a square root gives you a field extension of degree dividing 2.

For the second part we'll again use `extdgmul` and the fact that adjoining a cubic root of 2 gives an extension of Q of degree 3 (and similarly for trisecting the angle). What do you think?

**Comment by savask** [comment]
Since `fldGen` was merged into set.mm, maybe it's time to move on and see what we can do with it. Spoiler - it seems that we need field towers after all.

I'll use the definition of `Constr` from https://github.com/metamath/set.mm/issues/4246#issuecomment-2574581332 Now, let's use the following, shorter definition of `ConstrAlg` (later we'll see that it will have to be redefined, but bear with me for now):
```
df-constralg $a |- ConstrAlg = U. { f e. ~P CC | ( ( CCfld |`s f ) /FldExt ( CCfld |`s QQ ) /\ E. k e. NN ( ( CCfld |` f ) [:] ( CCfld |`s QQ ) ) = ( 2 ^ k ) ) } $.
```
Here `f` is a subset of `CC` which gives a subfield of C, and the degree |f : Q| is a power of two.

Let's use `F` for all such subsets of `CC`:
```
2flds $e |- F = { f e. ~P CC | ( ( CCfld |`s f ) /FldExt ( CCfld |`s QQ ) /\ E. k e. NN ( ( CCfld |`s f ) [:] ( CCfld |`s QQ ) ) = ( 2 ^ k ) ) } $.
```
Adding a square root to a field `A` from the class `F` will give us a (potentially larger) field from `F`:
```
2fldssqr.a $e |- ( ph -> A e. F ) $.
2fldssqr.x $e |- ( ph -> X e. CC ) $.
2fldssqr.s $e |- ( ph -> ( X x. X ) e. A ) $.
2fldssqr $p |- ( ph -> ( CCfld fldGen ( A u. { X } ) ) e. F ) $= ? $.
```
This should be relatively easy to prove from the definition of field extension degree. Now, the following statement seems to be more problematic: given two fields `A` and `B` from `F`, their composite field is again in that class.
```
2fldscomp.a $e |- ( ph -> A e. F ) $.
2fldscomp.b $e |- ( ph -> B e. F ) $.
2fldscomp $p |- ( ph -> ( CCfld fldGen ( A u. B ) ) e. F ) $= ? $.
```
I think to prove this we need to redefine `ConstrAlg` to use towers :-( The idea is that for B there should exist a tower of intermediate subfields Q = B1, B2, ... Bk = B, such that |B_{i+1} : B_i| = 2, or, equivalently, B_{i+1} is generated by B_i and some square root x_i of an element from B_i. Then one can consider fields A(x_1), A(x_1, x_2), ..., A(x_1, ..., x_k), and derive that the corresponding degrees are at most 2. Now, if AB is the composite field then it quickly follows that |AB : Q| is a power of 2 (and in fact there's also a tower of required type).

Now, the rest of what I'm going to say still holds. The intersection `X` of two lines defined over the "good" field `A` also lies in a good field:
```
constr2fldlem1.a $e |- ( ph -> A e. F ) $.
constr2fldlem1.x $e |- ( ph -> X e. CC ) $.
constr2fldlem1.s $e |- ( ph -> E. a e. A E. b e. A E. c e. A E. d e. A E. t e. RR E. r e. RR ( X = ( a + ( t x. ( b - a ) ) ) /\ X = ( c + ( r x. ( d - c ) ) ) /\ -. ( Im ` ( ( * ` ( b - a ) ) x. ( d - c ) ) ) = 0 ) ) $.
constr2fldlem1 $p |- ( ph -> E. s e. F X e. s ) $= ? $.
```
An intersection of a line and a circle:
```
constr2fldlem2.a $e |- ( ph -> A e. F ) $.
constr2fldlem2.x $e |- ( ph -> X e. CC ) $.
constr2fldlem2.s $e |- ( ph -> E. a e. A E. b e. A E. c e. A E. d e. A E. e e. A E. f e. A E. t e. RR ( X = ( a + ( t x. ( b - a ) ) ) /\ ( abs ` ( X - c ) ) = ( abs ` ( e - f ) ) ) ) $.
constr2fldlem2 $p |- ( ph -> E. s e. F X e. s ) $= ? $.
```
And an intersection of two circles:
```
constr2fldlem3.a $e |- ( ph -> A e. F ) $.
constr2fldlem3.x $e |- ( ph -> X e. CC ) $.
constr2fldlem3.s $e |- ( ph -> E. a e. A E. b e. A E. c e. A E. d e. A E. e e. A E. f e. A E. t e. RR ( -. a = d /\ ( abs ` ( X - a ) ) = ( abs ` ( b - c ) ) /\ ( abs ` ( X - d ) ) = ( abs ` ( e - f ) ) ) ) $.
constr2fldlem3 $p |- ( ph -> E. s e. F X e. s ) $= ? $.
```
The proofs of these will probably require only some arithmetic and `2fldssqr`. As a result we should be able to prove
```
constr2fld $p |- Constr C_ ConstrAlg $= ? $.
```
Now, to prove the impossibility of doubling the cube, we will first show that adjoining a cubic root of 2 gives an extension of degree 3:
```
cr2fld.x $e |- ( ph -> X e. CC ) $.
cr2fld.c $e |- ( ph -> ( X ^c 3 ) = 2 ) $.
cr2fld $p |- ( ph -> ( ( CCfld |`s ( CCfld fldGen ( QQ u. { X } ) ) ) [:] ( CCfld |`s QQ ) ) = 3 ) $= ? $.
```
hence this root is not (algebraically) constructible (here we use multiplicativity of degrees):
```
cr2ncnstr $p |- ( ph -> -. X e. ConstrAlg ) $= ? $.
```
And finally the desired result:
```
cantdouble $p |- -. ( 2 ^c ( 1 / 3 ) ) e. Constr $= ? $.
```
All labels and statements are of course very preliminary.

So what do you think?
1. How do we define towers? You mentioned chains in posets, but now I'm thinking that maybe a sequence of fields (i.e. a map from some interval) is more appropriate.
2. Maybe you see a way to avoid towers?
3. Should we create some file in the set.mm repository like `constructible.mm` or even `problems.mm` to collect unproved statements? Once something is formalized, it can be moved into a mathbox.

**Comment by tirix** [comment]
It's really nice to see this slowing taking form! (though the way is still very long!)

As for the method:

> Should we create some file in the set.mm repository like constructible.mm or even problems.mm to collect unproved statements? Once something is formalized, it can be moved into a mathbox.

I don't know what's best, I don't think this was ever done before. 

At some point in time, there was a suggestion to have a kind of purgatory (between main's heaven and mathboxes hell !) where common work which was not quite ready for main could live. The advantage of staying in one file is that we benefit from CI and all kind of label renamings (there seem to be a lot lately). The advantage of having our own file is that we are not bound to CI, collaboration directly from the GitHub interface is possible and easy (e.g. merges).

I think Lean is working with different GitHub repositories, but they have a different approach altogether.

In any case this could be discussed more largely.

**Comment by tirix** [comment]
> How do we define towers? You mentioned chains in posets, but now I'm thinking that maybe a sequence of fields (i.e. a map from some interval) is more appropriate.

I believe towers could be defined from [`Word`](https://us.metamath.org/mpeuni/df-word.html); words are sequences of anything and we can add a condition. Interestingly, [`UpWord`](https://us.metamath.org/mpeuni/df-upword.html) is very recent and resembles the kind of ordering condition we need (but is limited to `RR`)

**Comment by tirix** [comment]
> Since ` fldGen ` was merged into set.mm, maybe it's time to move on and see what we can do with it. 

I've added a few theorems in #4574 , including that ` QQ ` is generated by ` 1 `. This could be part of the initial step of the induction, since we start with `{ 0 , 1 }` only, the first field will be `QQ`.

**Comment by savask** [comment]
> I don't know what's best, I don't think this was ever done before. ...

I think we can't do it in the main set.mm file, since CI won't accept unproved statements (as far as I understand that's the reason why many people keep "future theorems" commented out). I suggest we create a file "constructible.mm" starting with `$[ set.mm $]`  and containing all the unproved statements. Once some chunk is proved, it can be moved to a mathbox. It shouldn't create problems for other contributors, but if others refuse, we could keep the file elsewhere.

As for my own question
> Maybe you see a way to avoid towers?

The answer is apparently no, there are elements x in C such that |Q(x) : Q| = 2^k, yet x isn't constructible, so we really need towers after all.

> I believe towers could be defined from [Word](https://us.metamath.org/mpeuni/df-word.html); words are sequences of anything and we can add a condition. Interestingly, [UpWord](https://us.metamath.org/mpeuni/df-upword.html) is very recent and resembles the kind of ordering condition we need (but is limited to RR)

Okay, what about the following definition of algebraically constructible numbers:
```
df-constralg $a |- ConstrAlg = U. { f e. ~P CC | ( ( CCfld |`s f ) /FldExt ( CCfld |`s QQ ) /\ E. w e. Word ( SubDRing ` CCfld ) ( ( w ` 0 ) = QQ /\ ( w ` ( # ` w ) ) = f /\ A. k e. ( 0 ..^ ( ( # ` w ) - 1 ) ) ( ( CCfld |`s ( w ` ( k + 1 ) ) ) [:] ( CCfld |`s ( w ` k ) ) ) = 2 ) ) } $.
```
We take all subsets `f` of `CC`, such that it gives a field extension of `QQ`, and there exists a word `w` consisting of subfields of `CCfld`, where the first "letter" is `QQ` and the last is `f` itself, while any two adjacent letters give a field extension of degree 2. Does it look correct? If yes, then we more or less have an initial proof plan.

> I've added a few theorems in https://github.com/metamath/set.mm/pull/4574 , including that QQ is generated by 1. This could be part of the initial step of the induction, since we start with { 0 , 1 } only, the first field will be QQ.

Nice! Maybe we'll be able to avoid using this fact, since we should be fine with `Constr C_ ConstrAlg`, in particular, the initial step of the induction simply becomes `{ 0 , 1 } C_ QQ`.

**Comment by tirix** [comment]
> I suggest we create a file "constructible.mm"

Ok, let's try it and see how the community reacts! Feel free to launch a PR if you wish, otherwise I might do it in my next PR.

> Does it look correct? If yes, then we more or less have an initial proof plan.

Yes it looks correct, with the only difference that it should be ``( w ` ( ( # ` w ) - 1 ) ) = f`` (since we start at index 0, the last index is the length minus one). @avekens has made a nice shortcut for that, [`lastS`](https://us.metamath.org/mpeuni/df-lsw.html), which we could use here.

Globally this seems like a very long definition though, I think I (still!) would like to abstract out the "tower" part, since it should be reusable. I might try to make a proposal.

**Comment by savask** [comment]
> ... otherwise I might do it in my next PR.

Sure, I'll be happy if you do it, especially since you propose to abstract out the tower part. Feel free to change the labels etc as you wish.

**Comment by tirix** [comment]
Ok, I'm still working on the proof that the degree of an algebraic extension $F/a$ within a field $E$ is the degree of the minimal polynomial, but I'm slowly getting there.

I can prove that the mapping $G : p \in F[x] \mapsto p(a)$ is an homomorphism, so its kernel $K$ is a prime ideal generated by the minimal polynomial. That mapping induces a field homomorphism $J : F[x]/K \rightarrow F$ , I have already proven that it is injective, but now I also need to prove that it is surjective onto the field generated by $`F \cup \{ a \}`$ (to apply [`ghmqusker`](https://us.metamath.org/mpeuni/ghmqusker.html)).

All proofs I've seen, including [Lang], [lecture notes kindly provided by Gérard Lang](https://fr.scribd.com/document/421330707/Antoine-Chambert-Loir-Algebre-Lecture-Notes-2015), and [the stacks project](https://stacks.math.columbia.edu/tag/09G1) gloss over it: "_we see easily that the map is surjective_"

I need to prove `ran G = ( E fldGen ( F u. { A } ) )` .
* for one direction `( E fldGen ( F u. { A } ) ) C_ ran G`, I can prove that `( F u. { A } ) C_ ran G`, but then I need to show that `ran G` is a sub-division-ring, and I don't think that's the case.

I'll try using ring spans `RSpan`.

**Comment by savask** [comment]
Correct me if I misunderstood your setting: you have a field $E$ with a subfield $F$ and an element $a \in E$. You have a homomorphism $J : F[x]/K \to E$ (I believe you have a typo in your post), and you know that $F[x]/K$ is a field. But then the image of $J$ is also a field (I think that's theorem `rndrhmcl` from your recent pull request). Images of $G$ and $J$ coincide, so that should do the trick. Am I missing some subtlety?

**Comment by tirix** [comment]
That will work, thank you so much!
I probably need a few more utility theorems though, so you'll have to be more patient to see the result... 

What about the other direction? I also need `ran G C_ ( E fldGen ( F u. { A } ) )`.
I think I might try ``ran G C_ ( ( RSpan ` E ) ` ( F u. { A } ) )`` (the ideal generated by $`F \cup \{ a \}`$). And since `( E fldGen ( F u. { A } ) )` is a field, its only non trivial ideal is itself.

(you're right about the typo in my post, too)

**Comment by savask** [comment]
> What about the other direction?

Your polynomial $p$ has coefficients in $F$, in particular, it has coefficients in $F(a)$, or `( E fldGen ( F u. { A } ) )` in set.mm notation. When you substitute $a$ in $p$, you get an element of $F(a)$. I think set.mm's theorem [fveval1fvcl](https://us.metamath.org/mpeuni/fveval1fvcl.html) is relevant.

**Comment by tirix** [comment]
Yes, that's perfect, thanks again!

**Comment by tirix** [comment]
With #4911, I think I'm getting nearer to the famous towers of quadratic field extensions.
(For the record, recent discussions about this project can also be found in https://github.com/metamath/set.mm/pull/4903#discussion_r2159178287 and in https://github.com/metamath/set.mm/pull/4872#issuecomment-2943844939)

One possible issue I'm seeing is that new constructible numbers will be build using several others, which will all belong to some towers of quadratic extensions - but possibly different ones. 

So I think the next step will be to prove that if `X` and `Y` belong to (possibly different) towers of quadratic field extensions, any linear combination of `X` and `Y` also belongs to (in general a third) tower of quadratic field extensions.

Or is there any clever way around this?

**Comment by savask** [comment]
> One possible issue I'm seeing is that new constructible numbers will be build using several others, which will all belong to some towers of quadratic extensions - but possibly different ones.

The idea is to prove that if some $z \in \mathbb{C}$ is geometrically constructible, then there exists a sequence of points $z_0, z_1, \dots, z_n$, where $z_0 = 0$, $z_1 = 1$, $z_n = z$ and each $z_i$ is obtained from $z_0, \dots, z_{i-1}$ by line-line, line-circle or circle-circle intersections, where parameters of lines and circles are taken from $z_0, \dots, z_{i-1}$. This is more or less the definition of a geometrically constructible point.

I think right now the definition of `Constr` in `constr.mm` is a bit different, since it takes the set $\{0, 1\}$ and keeps enlarging it by ruler and compass constructions. The definition in terms of a sequence of points, that I gave in the previous paragraph, can be derived from our definition of `Constr` by induction on the number of iterations. For example, let's say some point $z$ is the intersection of two lines. Each line is given by two parameters, which must have been constructed on the previous iteration, hence there exist corresponding sequences of points. By concatenating all these sequences and appending $z$ at the end, we obtain a correct sequence for $z$.

Alternatively, you can just change the definition of `Constr` to be formulated in terms of sequences :-)

Once you have a sequence of points, your tower of fields will be $\mathbb{Q} \leq \mathbb{Q}(z_0) \leq \mathbb{Q}(z_0, z_1) \leq \dots \leq \mathbb{Q}(z_0, z_1, \dots, z_n)$. The fact that each extension is at most quadratic follows from some basic (but probably technically unpleasant) algebra, i.e. one will have to prove that intersection points satisfy some quadratic equation.

**Comment by tirix** [comment]
Yes! Maybe I could just prove just the implication:
`z` constructible by current definition $\Rightarrow$ existence of a sequence of points $z_o, ... z_n = z$ $\Rightarrow$ existence of a tower of fields.

BTW, in the current definition the "next set of points" does not include the "current set of points", I think I might have to add `( s u. ...)` so that 
```
  df-constr $a |- Constr = ( rec ( ( s e. _V |-> { x e. CC | ( E. a e. s E. b e. s ...
```
would become
```
  df-constr $a |- Constr = ( rec ( ( s e. _V |-> ( s u. { x e. CC | ( E. a e. s E. b e. s ...
```
This way, the previous step constructible points are trivially included in the next step.

**Comment by savask** [comment]
>  I think I might have to add ( s u. ...) so that .. 

Sure, though it would be nice to formulate theorems in a way that it won't be hard to remove this addition later. Stewart in the book you linked earlier in Definition 7.2 doesn't assume that. Indeed, if $x$ is some nonzero point from the set $S$ and $0 \in S$, then $x$ lies on the intersection of a line passing from $0$ to $x$ and a circle $|z| = |x|$, so $x$ is constructible from $S$. Similarly, $0$ lies on the intersection of that line and a circle $|z-x| = |x|$.

So this is an easy argument, but probably something one can avoid for now.

**Comment by tirix** [comment]
> So this is an easy argument, but probably something one can avoid for now.

Yes, it's "easy", but not trivial :)

**Comment by savask** [comment]
@tirix What do you think about going back to the drawing board, i.e. `constr.mm`? I think all the essential definitions are in place, but I'm a bit afraid that the proof might go astray, especially if one tries to follow Stewart's book word to word.

For example, in Definition 7.6 Stewart defines the _Pythagorean closure_ $\mathbb{Q}^{py}$ which he proves to be equal to the field of geometrically constructible numbers. Then, in Theorem 7.11, he shows that $\mathbb{Q}^{py}$ consists of numbers reachable by towers of quadratic extensions, and the proof in one direction is a bit weird in my opinion. He says that for a number from $\mathbb{Q}^{py}$ there is a certain tower _by definition_ of $\mathbb{Q}^{py}$, while the definition says nothing about any towers at all.

Perhaps the implied argument is this: an element from $\mathbb{Q}^{py}$ is geometrically constructible by Theorem 7.7, hence there exists a sequence of points obtained through ruler and compass constructions, and this sequence of points defines a tower of quadratic extensions. This line of thought doesn't use any properties of $\mathbb{Q}^{py}$ at all, and that means we can completely avoid defining $\mathbb{Q}^{py}$ in the `set.mm` formalization.

In your recent pull request you noticed that Stewart requires the complex conjugate to be constructible, because formulae for intersections use it. Does that mean if $z_0, \dots, z_n$ is a sequence of constructible points, then the tower of fields won't be

$$\mathbb{Q} \leq \mathbb{Q}(z_0) \leq \mathbb{Q}(z_0, z_1) \leq \dots \leq \mathbb{Q}(z_0, z_1, \dots, z_n)$$

but rather

$$\mathbb{Q} \leq \mathbb{Q}(z_0) \leq \mathbb{Q}(z_0, \bar{z_0}) \leq \mathbb{Q}(z_0, \bar{z_0}, z_1) \leq \mathbb{Q}(z_0, \bar{z_0}, z_1, \bar{z_1}) \leq \dots \leq \mathbb{Q}(z_0, \bar{z_0}, z_1, \bar{z_1}, \dots, z_n, \bar{z_n}), $$

i.e. we extend the tower by adding a point and then its complex conjugate. The idea is to show that if $z$ is geometrically constructible, then there's a tower of quadratic extensions starting at $\mathbb{Q}$ and ending at $F$ containing $z$, such that $F$ is closed with respect to complex conjugation. I think this guarantees that all the coefficients we need for line-line, line-circle, circle-circle intersections lie in the subfields we want.

That being said, I'm not 100% sure that this approach is correct, so I think it would be good to draft theorem statements in `constr.mm` or here and see if anything unexpected pops up. What do you think?

**Comment by tirix** [comment]
Noting that each step of the construction only adds a finite number of new points, maybe we could actually still have a simpler formulation, with each step of the construction shown to be inside of the last field of a (single) tower of quadratic field extensions?

I'm thinking maybe the following:
1. prove that each step of the construction of constructible numbers has only a finite number of points, and only adds a finite number of new points.
2. prove that each new point is the solution of some linear (line-line) or quadratic (line-circle, circle-circle) equation with coefficients in the previous constructible set.
3. prove that given a finite family of points, each lying in a (possibly different) quadratic extension of some field, there is a single tower of quadratic field extensions which last field contains them all.
4. combine everything to show that  each step of the construction is included in the last field of a single tower of quadratic field extensions
5. finally, any constructible number is an element of some of the constructible step, therefore lies in the last field of some tower of quadratic field extensions.

**Comment by savask** [comment]
> I'm thinking maybe the following: ...

Yes, this plan should work, I was just hoping that there's some slick proof which will save formalization effort. If you're willing to prove more results from field/Galois theory, then what about the following idea.

First, one needs some results about field composites. If $E, F$ are subfields of a field $K$, then their composite $EF$ is the smallest subfield of $K$ containing $E$ and $F$. We don't need a new definition for composites, since it's just `K fldGen ( E u. F )` in `set.mm` notation. One can prove that if $F_1, F_2, E$ are subfields of $K$ and $F_1 \leq F_2$, then $|EF_2 : EF_1| \leq |F_2 : F_1|$. Indeed, if $S$ is a basis of $F_2$ over $F_1$, then $F_2 = F_1(S)$ and $EF_2 = F_1(E \cup S) = (F_1(E))(S)$, while $EF_1 = F_1(E)$. Hence $EF_2 = (EF_1)(S)$ and $|EF_2 : EF_1| \leq |S| = |F_2 : F_1|$ follows.

One corollary of this is that we can now glue towers of quadratic extensions. Let's say $\mathbb{Q} = F_1 \leq \dots \leq F_n$ is one such tower, and $\mathbb{Q} = E_1 \leq \dots \leq E_m$ is another. Then the following is also a tower of quadratic extensions:

$$\mathbb{Q} = F_1 \leq \dots \leq F_n = F_nE_1 \leq F_nE_2 \leq \dots \leq F_nE_m.$$

I think this roughly corresponds to what you have in part 3 of your plan.

Now we can define the set $\mathcal{C}$ of _algebraically constructible numbers_ as a union of all subfields $F$ of $\mathbb{C}$, such that there is a tower of quadratic extensions from $\mathbb{Q}$ to $F$. By the property above, if $F$ and $E$ are two such subfields, then their composite $EF$ also satisfies the tower property, so $\mathcal{C}$ is a field. I'm not sure we need a special constant for $\mathcal{C}$ in `set.mm`, after all one can show that it's equal to `Constr`.

Finally, one needs to prove that if $z \in \mathcal{C}$, then $\sqrt{z} \in \mathcal{C}$, i.e. adjoining a square root gives a quadratic extension, and also $\bar{z} \in \mathcal{C}$. This latter property requires the fact that $\bar{} : \mathbb{C} \to \mathbb{C}$ is a field automorphism, so it maps subfields to subfields and preserves the extension degree. So in particular it maps towers of quadratic extensions to similar towers, so $\mathcal{C}$ is closed with respect to complex conjugation.

If you have some further Galois theory in mind, it might be helpful to prove this later bit in more generality. Let's say, if $F$ is a subfield of $K$ and $\sigma : K \to K$ is a field automorphism, then $F^\sigma$ is also a subfield of $K$. One can take a look at Lang for examples of statements like that.

So having proved all that, one finishes the impossibility proof as follows. Since the field $\mathcal{C}$ contains $\mathbb{Q}$ and is closed with respect to complex conjugation and taking square roots, it will also be closed with respect to line-line, line-circle and circle-circle intersections, as follows from the formulae in Stewart. This should quickly imply that $\mathcal{C}$ contains all geometrically constructible numbers.

I guess this plan has the benefit that some of its intermediate statements (composite field, automorphisms) will be useful for other parts of field theory. But then it might be overly ambitious 😅

**Comment by tirix** [comment]
Yes, I'm interested in Galois theory and studying the field automorphisms certainly goes into the right direction.

I'll probably start with expressing that all 3 types of constructible numbers are roots of some quadratic equations (my step 2 and your last paragraph), this is a requirement in all methods anyway. Once I'm done I might have thought more about the next step.

I prefer smaller, more general theorems to big ad-hoc proofs, so your last proposal sounds very good. But I feel it might be a longer way.

Feel free to propose changes to `constr.mm` for example, it is a small file, it should even be editable in the browser using GitHub.

**Comment by tirix** [comment]
Ok, so #4926 has all the complicated formulas out of the way.

I think next I will work your proposal with composites for gluing towers of field extensions.

If I'm not mistaken, we will not need the automorphism part for the impossibility proofs stricto sensu, since we would only need the other inclusion, and not equality... which means we should be very near to the end!

**Comment by tirix** [comment]
> First, one needs some results about field composites. If $E, F$ are subfields of a field $K$, then their composite $EF$ is the smallest subfield of $K$ containing $E$ and $F$. We don't need a new definition for composites, since it's just `K fldGen ( E u. F )` in `set.mm` notation. One can prove that if $F_1, F_2, E$ are subfields of $K$ and $F_1 \leq F_2$, then $|EF_2 : EF_1| \leq |F_2 : F_1|$. Indeed, if $S$ is a basis of $F_2$ over $F_1$, then $F_2 = F_1(S)$ and $EF_2 = F_1(E \cup S) = (F_1(E))(S)$, while $EF_1 = F_1(E)$. Hence $EF_2 = (EF_1)(S)$ and $|EF_2 : EF_1| \leq |S| = |F_2 : F_1|$ follows.

Ok, I've tried to formalize this argument, but I think I'm missing something.  I've been looking at [proposition 116 p.32 of lecture notes I found online](https://www3.nd.edu/~ajorza/courses/2015s-m60220/notes/lectures.pdf), which I think is relevant here. Following these notes, let's assume $K$ is a subfield of both $L_1$ and $L_2$. The proof provided says $L_1 L_2 = K(u_i, v_j)$ where $u_i$ and $v_j$ are basis of $L_1/K$ and $L_2/K$ respectively. If I take $A \in L_1 \textbackslash K$ and $B \in L_2 \textbackslash K$, then $\frac{1}{A+B} \in L_1 L_2$, but I don't see how $\frac{1}{A+B} \in K(u_i, v_j)$, which only contains _linear_ combinations...

**Comment by savask** [comment]
> which only contains linear combinations...

Here notation $K(u_i, v_j)$ means the smallest field generated by $K$ and $u_i, v_j$ (i.e. `fldGen` in set.mm notation), so it contains fractions. But then it's not clear why $K(u_i, v_j)$ is (linearly) spanned by $u_1, \dots, u_m$ over $L_2$.

I got confused by this in the argument which you cite, so I assumed $EF_2 = (EF_1)(S)$ implies $|EF_2 : EF_1| \leq |S|$, but it's not true in general that $|F(S) : F| \leq |S|$. Say, if $S$ is a singleton, then it would imply $|F(S) : F| \leq 1$, which is clearly false for elements of large degree over $F$. Sorry for wasting your time :-(

The trick here is the following theorem (Bourbaki, Algebra 2, Chapter V, Proposition 1):

<img width="743" height="205" alt="Image" src="https://github.com/user-attachments/assets/562b14d1-2973-4269-acc0-05b4a49e9933" />

In the notation of the lectures you cite, let's take $A = K[L_1 \cap L_2]$, i.e. a _subring_ (not a subfield) generated by $L_1$ and $L_2$. It consists of all linear combinations of the form $\alpha_1 \beta_1 + \dots + \alpha_k \beta_k$, where $\alpha_i \in L_1$ and $\beta_i \in L_2$. Since $L_2$ is linearly spanned by $v_1, \dots, v_n$ over $K$, the ring $A$ actually consists of linear combinations of the form $\alpha_1 v_1 + \dots + \alpha_n v_n$, where $\alpha_i \in L_1$. In particular, $A$ is finite-dimensional over $L_1$, and it is an integral domain since it lives inside some larger field, hence by corollary above, $A$ is a field. This implies $A = L_1L_2$ at once.

To sum up, we've shown that $A$ is linearly spanned by $v_1, \dots, v_n$ over $L_1$, hence $|A : L_1| \leq n$. At the same time, $A = L_1L_2$, so $|L_1 L_2 : K| \leq |L_1 L_2 : L_1| \cdot |L_1 : K| \leq nm$.

Here's the relevant argument from Bourbaki, Algebra 2, Chapter V, Proposition 5:

<img width="745" height="410" alt="Image" src="https://github.com/user-attachments/assets/341e505d-e23d-4c4a-9e8a-c288f101dbf0" />

Feel free to ask for clarifications.

**Comment by tirix** [comment]
Thank you so much! 
This should unblock me!

**Comment by avekens** [comment]
I noticed that the proof for Metamath 100 #8 is still not mentioned in "Formalizing 100 Theorems" ([https://www.cs.ru.nl/~freek/100/](https://www.cs.ru.nl/%7Efreek/100/)). @david-a-wheeler could you inform Freek to add it, please?

**Comment by david-a-wheeler** [comment]
@avekens - I've nagged Freek to ask him to add it :-).
